iroha_macro/iroha_derive/src/lib.rs

```diff
@@ -155,6 +161,93 @@ fn impl_io(ast: &syn::DeriveInput) -> TokenStream {
     gen.into()
 }
 
+fn get_struct_arg<'a, 'b>(s: &'a str, ty: &'b syn::TypePath) -> Option<&'b syn::GenericArgument> {
+    let segs = &ty.path.segments;
+    if segs.len() != 1 || segs[0].ident != s {
+        return None;
+    }
+
+    if let syn::PathArguments::AngleBracketed(ref bracargs) = segs[0].arguments {
+        assert_eq!(bracargs.args.len(), 1);
+        Some(&bracargs.args[0])
+    } else {
+        unimplemented!()
+    }
+}
+
+fn from_boxed_variant_internal(
+    into_ty: &syn::Ident,
+    from_variant: &syn::Ident,
+    from_ty: &syn::GenericArgument,
+) -> proc_macro2::TokenStream {
+    quote! {
+        impl std::convert::From<#from_ty> for #into_ty {
+            fn from(origin: #from_ty) -> Self {
+                #into_ty :: # from_variant (Box::new(origin))
+            }
+        }
+    }
+}
+
+fn from_variant_internal(
+    into_ty: &syn::Ident,
+    from_variant: &syn::Ident,
+    from_ty: &syn::Type,
+) -> proc_macro2::TokenStream {
+    quote! {
+        impl std::convert::From<#from_ty> for #into_ty {
+            fn from(origin: #from_ty) -> Self {
+                #into_ty :: # from_variant (origin)
+            }
+        }
+    }
+}
+
+fn from_variant(
+    into_ty: &syn::Ident,
+    from_variant: &syn::Ident,
+    from_ty: &syn::Type,
+) -> proc_macro2::TokenStream {
+    let from_orig = from_variant_internal(into_ty, from_variant, from_ty);
+
+    if let syn::Type::Path(path) = from_ty {
+        if let Some(inner) = get_struct_arg("Box", path) {
+            let from_inner = from_boxed_variant_internal(into_ty, from_variant, inner);
+            return quote! {
+                #from_orig
+                #from_inner
+            };
+        }
+    }
+
+    from_orig
+}
+
+fn impl_from_variant(ast: &syn::DeriveInput) -> TokenStream {
+    let name = &ast.ident;
+
+    let froms = if let syn::Data::Enum(ref data_enum) = ast.data {
+        &data_enum.variants
+    } else {
+        unimplemented!()
```

Comment:
I would suggest to give some context, e.g. show a user `Only supports enums` or something like that.